feat: lite version of transaction information

zhangyouxin opened the pull request:
## Description

PRD: https://vs0cjf.axshare.com/?id=3giphq&p=_189_how_to_support_customers_and_developers_opera&g=1

Desktop design: https://www.figma.com/file/6XNoimRDbFTTNm016rbIdU/Magickbase?node-id=13042%3A37425&t=lriivPlPTOL3A8WN-1

Mobile design: https://www.figma.com/file/6XNoimRDbFTTNm016rbIdU/Magickbase?node-id=13058%3A41067&mode=dev

Resolve: https://github.com/Magickbase/ckb-explorer-public-issues/issues/189

Backend PR: https://github.com/nervosnetwork/ckb-explorer/pull/1384

Preview example: https://ckb-explorer-git-transaction-lite-magickbase.vercel.app/transaction/0x21b9d887b8a68bf56f1ae6bc7cd28d79395efd2241de3de57e6fdfd36714b2ee?layout=lite

Notes: 
 - the original PR is [#5](https://github.com/Magickbase/ckb-explorer-frontend/pull/5), but closed for work handover
 - cell tags are not implemented yet (eg: "mint", "nft transfer" tags)

Keith-CY commented on src/pages/Transaction/TransactionComp.tsx:
```diff
@@ -327,6 +357,80 @@ const handleCellbaseInputs = (inputs: State.Cell[], outputs: State.Cell[]) => {
+      {transactionLiteDetails &&
+        transactionLiteDetails.map((item, index) => {
+          return (
+            <div className="transaction_lite" key={index}>
```
Avoid using `index` as the `key`, what if using `item.address`

Ref:
- key of list: https://legacy.reactjs.org/docs/lists-and-keys.html#keys

Keith-CY commented on src/pages/Transaction/TransactionComp.tsx:
```diff
@@ -327,6 +357,80 @@ const handleCellbaseInputs = (inputs: State.Cell[], outputs: State.Cell[]) => {
+                <div className={styles.transactionLiteBoxContent}>
+                  {item.transfers.map((items, innerIndex) => {
+                    return (
+                      <div key={innerIndex}>
```
Avoid using `innerIndex` as the `key`, what if using `id` of the item`

Keith-CY commented on src/pages/Transaction/TransactionComp.tsx:
```diff
@@ -327,6 +357,80 @@ const handleCellbaseInputs = (inputs: State.Cell[], outputs: State.Cell[]) => {
+                  </div>
+                </div>
+                <div className={styles.transactionLiteBoxContent}>
+                  {item.transfers.map((items, innerIndex) => {
```
`items` here is confusing

Keith-CY commented on src/pages/Transaction/TransactionComp.tsx:
```diff
@@ -327,6 +357,80 @@ const handleCellbaseInputs = (inputs: State.Cell[], outputs: State.Cell[]) => {
+                            <span className={styles.nftId}>-ID : {items.nftId}</span>
+                          ) : (
+                            <>
+                              <span className={items.capacity > 0 ? styles.add : styles.subtraction}>
+                                {items.capacity > 0 ? '+' : ''}
+                              </span>
+                              <DecimalCapacity
+                                color={items.capacity > 0 ? '#00CC9B' : '#FA504F'}
+                                value={localeNumberString(shannonToCkb(items.capacity))}
+                              />
```
`Color` and `prefix` can be done by CSS to simplify the component

Ref:
- pseudo element: https://stackoverflow.com/questions/16392412/how-to-prefix-this-input-element-using-css

zhangyouxin commented:
Please have a review @Keith-CY , The comments here are mostly on legacy codes, some of them are not very maintainable, I just refactored some for this PR. What if we remove `styled` and `BEM`s? just by using `SCSS` could simplify the style sheets, I can gradually do the refactor in future development

Keith-CY commented:
> Please have a review @Keith-CY , The comments here are mostly on legacy codes, some of them are not very maintainable, I just refactored some for this PR. What if we remove `styled` and `BEM`s? just by using `SCSS` could simplify the style sheets, I can gradually do the refactor in future development

Sure, removing `styled-component` is one of our goals, so do `BEM`

Keith-CY commented on src/index.css:
```diff
@@ -2,6 +2,8 @@
 
 body {
   --primary-color: #00cc9b;
+  --emphasis-color: #FA504F;
```
It's usually named `accent color`

Ref:
- accent color: https://www.thespruce.com/what-are-accent-colors-5218695

zhangyouxin replied:
updated, learned a lot here, I thought accent color was a set of color pallet.

Keith-CY commented on src/pages/Transaction/styles.module.scss:
```diff
@@ -0,0 +1,158 @@
+      height: auto;
+    }
+  }
+}
```
```suggestion
}

```

Keith-CY commented on src/pages/Transaction/TransactionComp.tsx:
```diff
@@ -327,6 +331,78 @@ const handleCellbaseInputs = (inputs: State.Cell[], outputs: State.Cell[]) => {
+              <div className={styles.transactionLiteBoxContent}>
+                {item.transfers.map((transfer, index) => {
+                  return (
+                    <div key={`transfer-${index}`}>
```
It still relies on `index`. Does `transfer` has its own id, e.g. `tokenId`. I see `transfer.nftId` is mentioned below

zhangyouxin replied:
I think none of its attribute is unique: https://github.com/Magickbase/ckb-explorer-frontend/blob/transaction-lite/src/types/index.d.ts#L140-L152

Keith-CY replied:
I see

WhiteMinds commented on src/components/Card/HashCard/index.tsx:
```diff
@@ -120,7 +133,7 @@ export default ({
               {!loading && <img src={CopyIcon} alt="copy" />}
             </SimpleButton>
-            {counterpartAddr ? (
+            {counterpartAddr && (
```
> Do we have a good reason to use a short circuit expression instead of the ternary?

This is a syntax sugar for the pattern of unable to directly write `if (bool) return node` in jsx.
React will ignore ReactNode that is a boolean value.
I prefer this approach because it is widespread, similar to `[...arr]`.

Keith-CY replied:
> > Do we have a good reason to use a short circuit expression instead of the ternary?
> 
> This is a syntax sugar for the pattern of unable to directly write `if (bool) return node` in jsx. React will ignore ReactNode that is a boolean value. I prefer this approach because it is widespread, similar to `[...arr]`.

It has an implicit precondition that `condition` should be the type of `boolean`. So it's prone to render other falsy values, e,g. `0`, `NaN`

Keith-CY replied:
> > > Do we have a good reason to use a short circuit expression instead of the ternary?
> > 
> > 
> > This is a syntax sugar for the pattern of unable to directly write `if (bool) return node` in jsx. React will ignore ReactNode that is a boolean value. I prefer this approach because it is widespread, similar to `[...arr]`.
> 
> It has an implicit precondition that `condition` should be the type of `boolean`. So it's prone to render other falsy values, e,g. `0`, `NaN`

I found it was highlighted in react docs that `not to put numbers on the left side of &&`

![image](https://github.com/Magickbase/ckb-explorer-frontend/assets/7271329/e6cb9c38-fb40-45c0-a850-46d49b56e002)

Ref: https://react.dev/learn/conditional-rendering

WhiteMinds replied:
> It has an implicit precondition that `condition` should be the type of `boolean`. So it's prone to render other falsy values, e,g. `0`, `NaN`

> I found it was highlighted in react docs that `not to put numbers on the left side of &&`

If the developer is not aware of this issue, using ternary syntax or using the `&&` syntax will have the same result.
If they are aware of this issue, the result will still be the same.

Example:
```
const val: null | number = 0
return val && <i>val</i>
return val ? <i>val</i> : null
// correct
return val != null && <i>val</i>
return val != null ? <i>val</i> : null
```

Keith-CY replied:
> > It has an implicit precondition that `condition` should be the type of `boolean`. So it's prone to render other falsy values, e,g. `0`, `NaN`
> 
> > I found it was highlighted in react docs that `not to put numbers on the left side of &&`
> 
> If the developer is not aware of this issue, using ternary syntax or using the `&&` syntax will have the same result. If they are aware of this issue, the result will still be the same.
> 
> Example:
> 
> ```
> const val: null | number = 0
> return val && <i>val</i>
> return val ? <i>val</i> : null
> // correct
> return val != null && <i>val</i>
> return val != null ? <i>val</i> : null
> ```

I didn't get the example

```
const val: null | number = 0
return val && <i>val</i> // => render 0
return val ? <i>val</i> : null // => render nothing
```

Keith-CY replied:
> > > It has an implicit precondition that `condition` should be the type of `boolean`. So it's prone to render other falsy values, e,g. `0`, `NaN`
> > 
> > 
> > > I found it was highlighted in react docs that `not to put numbers on the left side of &&`
> > 
> > 
> > If the developer is not aware of this issue, using ternary syntax or using the `&&` syntax will have the same result. If they are aware of this issue, the result will still be the same.
> > Example:
> > ```
> > const val: null | number = 0
> > return val && <i>val</i>
> > return val ? <i>val</i> : null
> > // correct
> > return val != null && <i>val</i>
> > return val != null ? <i>val</i> : null
> > ```
> 
> I didn't get the example
> 
> ```
> const val: null | number = 0
> return val && <i>val</i> // => render 0
> return val ? <i>val</i> : null // => render nothing
> ```

Do you mean `rendering 0` is expected, but `val ? <i>val</i> : null renders nothing`, which is unexpected.

In this case, the `falsy` branch should be `0` instead of `null` because it's expected to render something, `null` is unexpected.

The main advantage the ternary brings is handling unexpected `condition`

WhiteMinds replied:
> return val && <i>val</i> // => render 0

Oh, I made a rookie mistake and forgot that this expression will return val. In that case, the ternary syntax should indeed be used first.

WhiteMinds commented on src/components/Card/HashCard/index.tsx:
```diff
@@ -51,12 +52,25 @@ export default ({
+  const updateSearchParams = useUpdateSearchParams<'layout'>()
+  const layout = searchParams.layout === Lite ? Lite : defaultLayout
+
+  const onChangeLayout = (layoutType: LayoutLiteProfessional.Professional | LayoutLiteProfessional.Lite) => {
```
```suggestion
  const onChangeLayout = (layoutType: LayoutLiteProfessional) => {
```

WhiteMinds commented on src/pages/Address/AddressComp.tsx:
```diff
@@ -313,17 +314,18 @@ export const AddressTransactions = ({
   const totalPages = Math.ceil(total / pageSize)
 
-  const onChangeLayout = (lo: 'professional' | 'lite') => {
+  const onChangeLayout = (layoutType: LayoutLiteProfessional.Professional | LayoutLiteProfessional.Lite) => {
```
```suggestion
  const onChangeLayout = (layoutType: LayoutLiteProfessional) => {
```

WhiteMinds commented on src/pages/Transaction/index.tsx:
```diff
@@ -24,20 +27,31 @@ export default () => {
   const transaction = query.data ?? defaultTransactionInfo
   const { blockTimestamp, txStatus } = transaction
+  const searchParams = useSearchParams('layout')
+  const layout: string = searchParams.layout === Lite ? Lite : Professional
```
```suggestion
  const layout = searchParams.layout === Lite ? Lite : Professional
```

WhiteMinds commented on src/pages/Transaction/TransactionComp.tsx:
```diff
@@ -88,7 +95,10 @@ const TransactionInfoItemWrapper = ({
 )
 
-export const TransactionOverview: FC<{ transaction: State.Transaction }> = ({ transaction }) => {
+export const TransactionOverview: FC<{ transaction: State.Transaction; layout: string }> = ({
```
```suggestion
export const TransactionOverview: FC<{ transaction: State.Transaction; layout: LayoutLiteProfessional }> = ({
```

Keith-CY commented:
Please have a review @FrederLu @silySuper 
And please resolve the conflicts @zhangyouxin

FrederLu commented:
1、Information such as CKB is missing under the address on the page.
URL  https://ckb-explorer-4rek3x5hw-magickbase.vercel.app/transaction/0xa60274bbf76d6f9d5c507335a09ba7a5044bef9cea170578d64f6a03ae3227f3
UI   https://www.figma.com/file/6XNoimRDbFTTNm016rbIdU/Magickbase?type=design&node-id=13058-39866&mode=design&t=3yAFERidb61M1zrX-0
<img width="1261" alt="image" src="https://github.com/Magickbase/ckb-explorer-frontend/assets/25676576/cc1d78f0-ff35-4e33-82f2-9ad107b05531">
<img width="1292" alt="image" src="https://github.com/Magickbase/ckb-explorer-frontend/assets/25676576/38d1a3d3-2e9a-4cd2-a659-354770e75f13">

---
2、Nervos DAO action tags are missing in Lite. At the same time, there must be a prompt message when the mouse hovers.
URL https://ckb-explorer-4rek3x5hw-magickbase.vercel.app/transaction/0x46e8c6fd3e8326b25d507b146cb715ecbf63c5933d575c1ed8e701a29a3ff4ba
<img width="1260" alt="image" src="https://github.com/Magickbase/ckb-explorer-frontend/assets/25676576/733e229c-3b89-4e51-8001-cfda4af580e6">
<img width="1297" alt="image" src="https://github.com/Magickbase/ckb-explorer-frontend/assets/25676576/a222effb-2481-4e61-8f46-37e61da88065">
<img width="844" alt="image" src="https://github.com/Magickbase/ckb-explorer-frontend/assets/25676576/ce09823f-50de-4496-9769-9359655ce7d0">


---
3、When `Nervos DAO Deposit/Nervos DAO Withdraw` appears in Professional for the same transaction and the same address, you can click the icon on the list page to view the display data to determine whether `Deposit/Withdraw` is displayed.
URL https://ckb-explorer-4rek3x5hw-magickbase.vercel.app/transaction/0x63a9653f593f243dfddee8075877d1f6e074946a39d9a720b2425fd05290d783


https://github.com/Magickbase/ckb-explorer-frontend/assets/25676576/920a528d-2ce1-4314-a407-c48814da9499

Keith-CY commented:
Any update for the feedback at https://github.com/Magickbase/ckb-explorer-frontend/pull/25#issuecomment-1707940472

Keith-CY commented:
How is it going

FrederLu commented:
Some of the issues have been addressed, but the whole thing is still being worked on. 
@zhangyouxin You can synchronize some information.

zhangyouxin commented:
For the comments:

1,3,4. the design is being adjusted @Sven-TBD 
2. added
5. the `api/v2/das_accounts` api for address `ckt1qqexmutxu0c2jq9q4msy8cc6fh4q7q02xvr7dc347zw3ks3qka0m6qgres90ptu3rh2qs5acer07ay9n97x3ajkkq0xq4u90jywagzznhrydlm5skvhc68k26cucdqcn` on staging env didn’t bind an alias @rabbitz 
6. fixed
7. fixed

FrederLu commented:
> For the comments:
> 
> 1,3,4. the design is being adjusted @Sven-TBD 2. added 5. the `api/v2/das_accounts` api for address `ckt1qqexmutxu0c2jq9q4msy8cc6fh4q7q02xvr7dc347zw3ks3qka0m6qgres90ptu3rh2qs5acer07ay9n97x3ajkkq0xq4u90jywagzznhrydlm5skvhc68k26cucdqcn` on staging env didn’t bind an alias @rabbitz 6. fixed 7. fixed

6, and 7 have been verified, others need to be adjusted.

The label at point 2 has been added, but the size is too large, affecting the font size later.
<img width="931" alt="image" src="https://github.com/Magickbase/ckb-explorer-frontend/assets/25676576/ff361ddc-00bc-4946-9489-d0daab08309a">
<img width="1264" alt="image" src="https://github.com/Magickbase/ckb-explorer-frontend/assets/25676576/e1eef6e5-50ff-46d9-876e-d90b2329e164">

Sven-TBD commented:
The UI link has been updated, still a few points should be aware of:

- For each different Asset type, there will be a line.
- Only for DAO-related transaction, a tag prompt is required

rabbitz commented:
> For the comments:
> 
> 1,3,4. the design is being adjusted @Sven-TBD 2. added 5. the `api/v2/das_accounts` api for address `ckt1qqexmutxu0c2jq9q4msy8cc6fh4q7q02xvr7dc347zw3ks3qka0m6qgres90ptu3rh2qs5acer07ay9n97x3ajkkq0xq4u90jywagzznhrydlm5skvhc68k26cucdqcn` on staging env didn’t bind an alias @rabbitz 6. fixed 7. fixed

```bash
root@ckb-explorer-api-6ccbcb8d9d-zgx76:/usr/src# curl -X POST http://das-account-indexer:8122/v1/reverse/record -d'{"type": "blockchain","key_info":{"coin_type": "", "chain_id": "1", "key": "0xcc0af0af911dd40853b8c8dfee90b32f8d1ecad6"}}'
{"err_no":0,"err_msg":"","data":{"account":"","account_alias":""}}
```

There are no available domains for this address (testnet).

zhangyouxin commented:
> The UI [link](https://www.figma.com/file/6XNoimRDbFTTNm016rbIdU/Magickbase?type=design&node-id=13058-39866&mode=design&t=JrIUzeZxQHa0UxtQ-4) has been updated, still a few points should be aware of:
> 
> * For each different Asset type, there will be a line.
> * Only for DAO-related transaction, a tag prompt is required

@rabbitz please check this out, there should be some changes for the api

zhangyouxin commented:
> The UI [link](https://www.figma.com/file/6XNoimRDbFTTNm016rbIdU/Magickbase?type=design&node-id=13058-39866&mode=design&t=JrIUzeZxQHa0UxtQ-4) has been updated, still a few points should be aware of:
> 
> * For each different Asset type, there will be a line.
> * Only for DAO-related transaction, a tag prompt is required

@rabbitz please check this out, there should be some changes for the api

zhangyouxin commented:
@rabbitz  According to the new design, the `sUDT` and `NFT` transfers would contain capacity ocupation amount, `NFT` transfers would show nft id:
<img width="962" alt="image" src="https://github.com/Magickbase/ckb-explorer-frontend/assets/3870972/2718a15e-24ab-4b0a-a7b3-069c85ef31ea">

FrederLu commented:
> @rabbitz According to the new design, the `sUDT` and `NFT` transfers would contain capacity ocupation amount, `NFT` transfers would show nft id: <img alt="image" width="962" src="https://user-images.githubusercontent.com/3870972/270207530-2718a15e-24ab-4b0a-a7b3-069c85ef31ea.png">

@Sven-TBD 
I saw `()` in the number part in the screenshot. Where did the value `61CKB` come from?

rabbitz commented:
related pr: https://github.com/nervosnetwork/ckb-explorer/pull/1463

FrederLu commented:
1、In UDT transactions, `+/—` matches and displays abnormally, and the `capacity` value displays abnormally.
https://ckb-explorer-frontend-in-magickbase-repo-jm1nw532y-magickbase.vercel.app/transaction/0xd8668c851f09559868d4b91340670c4ee9a6dcf50184019cf42dcd8fa39a745b?layout=lite
<img width="1317" alt="image" src="https://github.com/Magickbase/ckb-explorer-frontend/assets/25676576/4bde0a19-cd97-48dc-8d1a-1d3922f6a471">

---
2、The layout on the mobile terminal needs to be adjusted with reference to the UI.
<img width="1189" alt="image" src="https://github.com/Magickbase/ckb-explorer-frontend/assets/25676576/966e413d-bfd1-45ce-8633-938512307dd6">

<img width="312" alt="image" src="https://github.com/Magickbase/ckb-explorer-frontend/assets/25676576/7b024850-1984-4493-8dae-63205553a786">

FrederLu commented:
The page needs to be displayed in real time based on the api data, and the `-` symbol is currently missing.
https://ckb-explorer-frontend-in-magickbase-repo-12npn9j73-magickbase.vercel.app/transaction/0x6a98ccfa27fe375febd3648ac4bab9984da64de8c36f8268454c5cb6b79f6969?layout=lite
<img width="1425" alt="image" src="https://github.com/Magickbase/ckb-explorer-frontend/assets/25676576/ec771486-7031-41aa-a150-a2a1daac8c1e">

FrederLu commented:
@zhangyouxin 
Referring to the interface data, if cell_type: "m_nft_class", CKB needs to be displayed as `m-NFT-class` on the page.
https://ckb-explorer-frontend-in-magickbase-repo-5q0j3pbmh-magickbase.vercel.app/transaction/0x3752b2a5ecb7a05e1fbb6ed89f2d5df4db991c957975703883daf4739a78d918?layout=lite
<img width="1439" alt="image" src="https://github.com/Magickbase/ckb-explorer-frontend/assets/25676576/dce8a4ba-0b30-4ef8-a620-6ca171226bd9">
